# Actvidad 2 Open Data: Áreas infantiles municipales

A través de los datos abiertos del ayuntamiento de Madrid, exploraremos la base de datos de las Áreas infatiles municipales en Madrid del 2025 [link](https://datos.madrid.es/portal/site/egob/menuitem.c05c1f754a33a9fbe4b2e4b284f1a5a0/?vgnextoid=337541000be13410VgnVCM2000000c205a0aRCRD&vgnextchannel=374512b9ace9f310VgnVCM100000171f5a0aRCRD&vgnextfmt=default) para ver que información podemos extraer al respecto.

## 1. Importación de la base de datos y print para hacernos una idea de la estructura de los datos.


In [3]:
import pandas as pd
df_rawdata = pd.read_csv("https://raw.githubusercontent.com/bintutr/Data-Integration/refs/heads/main/areas_infantiles202501.csv", delimiter=';')

In [5]:
print(df_rawdata)

           ID DESC_CLASIFICACION  COD_BARRIO         BARRIO  COD_DISTRITO  \
0     5240286      Área Infantil         101   LOS CARMENES            10   
1       17031      Área Infantil          87     MIRASIERRA             8   
2       16666      Área Infantil          14       JUSTICIA             1   
3     5237133      Área Infantil         166   VALDEFUENTES            16   
4       52225      Área Infantil         192  VALDEBERNARDO            19   
...       ...                ...         ...            ...           ...   
2157    18277      Área Infantil         175        ANGELES            17   
2158    18118      Área Infantil         161        PALOMAS            16   
2159  9424149      Área Infantil          22        ACACIAS             2   
2160    17453      Área Infantil         117       ABRANTES            11   
2161  8914012      Área Infantil         143     MARROQUINA            14   

                   DISTRITO     ESTADO  COORD_GIS_X  COORD_GIS_Y  \
0      

Podemos observar que el Dataframe está compuesto por 23 columnas y 2162 filas. Así mismo, se observa que la existencia de valores nulos en algunas de las columnas.

# 2. En caso de no hacer un print del Dataframe, podríamos utilizar el código SHAPE para ver directamente las filas y columnas.

In [6]:
df_rawdata.shape

(2162, 23)

# 3. Adentrémonos un poco más a las características de los datos...

In [7]:
df_rawdata.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2162 entries, 0 to 2161
Data columns (total 23 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  2162 non-null   int64  
 1   DESC_CLASIFICACION  2162 non-null   object 
 2   COD_BARRIO          2162 non-null   int64  
 3   BARRIO              2162 non-null   object 
 4   COD_DISTRITO        2162 non-null   int64  
 5   DISTRITO            2162 non-null   object 
 6   ESTADO              2162 non-null   object 
 7   COORD_GIS_X         2162 non-null   float64
 8   COORD_GIS_Y         2162 non-null   float64
 9   SISTEMA_COORD       2162 non-null   object 
 10  LATITUD             2162 non-null   float64
 11  LONGITUD            2162 non-null   float64
 12  TIPO_VIA            2065 non-null   object 
 13  NOM_VIA             2065 non-null   object 
 14  NUM_VIA             2064 non-null   object 
 15  COD_POSTAL          2125 non-null   float64
 16  DIRECC

A través del código info, podemos observar el número total de filas que no tienen valores nulos, así mismmo, el tipo de datos que contiene cada una de las 22 columnas que conforman el Dataframe. Por último, se puede observar cuanta memoria ocupa el Dataframe.


# 3. Resumiendo el punto 2 con el código dtypes:

In [8]:
df_rawdata.dtypes

ID                      int64
DESC_CLASIFICACION     object
COD_BARRIO              int64
BARRIO                 object
COD_DISTRITO            int64
DISTRITO               object
ESTADO                 object
COORD_GIS_X           float64
COORD_GIS_Y           float64
SISTEMA_COORD          object
LATITUD               float64
LONGITUD              float64
TIPO_VIA               object
NOM_VIA                object
NUM_VIA                object
COD_POSTAL            float64
DIRECCION_AUX          object
NDP                   float64
FECHA_INSTALACION      object
CODIGO_INTERNO         object
CONTRATO_COD           object
TIPO                   object
TOTAL_ELEM              int64
dtype: object

En caso de querer obtener únicamente información con respecto al tipo de datos de cada columnna este código nos permitiria obtener únicamente está información.

# 4. Resumen estadístico de los datos:

In [9]:
df_rawdata.describe()

ID   COD_BARRIO  COD_DISTRITO    COORD_GIS_X   COORD_GIS_Y  \
count  2.162000e+03  2162.000000   2162.000000    2162.000000  2.162000e+03   
mean   7.477009e+05   128.978261     12.511101  442676.381984  4.474820e+06   
std    1.978134e+06    52.526021      5.253070    4328.329425  4.757706e+03   
min    1.665700e+04    11.000000      1.000000  429551.575000  4.464949e+06   
25%    1.732525e+04    93.000000      9.000000  439564.116750  4.470773e+06   
50%    1.791050e+04   133.000000     13.000000  443035.219000  4.474055e+06   
75%    1.849775e+04   166.000000     16.000000  446199.897500  4.478934e+06   
max    9.424149e+06   215.000000     21.000000  453385.843000  4.487423e+06   

           LATITUD     LONGITUD    COD_POSTAL           NDP   TOTAL_ELEM  
count  2162.000000  2162.000000   2125.000000  2.036000e+03  2162.000000  
mean     40.422025    -3.675724  27847.597647  1.941960e+07    10.548104  
std       0.042916     0.050942   2268.389082  8.429462e+06     5.509468  
min      40.332965    -3.831082      0.000000  1.100019e+07     0.000000  
25%      40.385558    -3.712255  28022.000000  1.109285e+07     7.000000  
50%      40.415158    -3.671464  28032.000000  2.006189e+07     9.000000  
75%      40.459122    -3.634144  28042.000000  3.101029e+07    13.000000  
max      40.534872    -3.549355  28055.000000  3.106697e+07    57.000000

A través de este resumen estadístico, podemos osbervar un resumen de las columnas númericas de nuestro Dataframe. Este resumen nos permite tener una idea rápida de la distribución de los datos de nuestro Dataframe.

# 5. Valores nulos

In [10]:
valores_nulos = df_rawdata.isnull().sum()
print(valores_nulos)

ID                      0
DESC_CLASIFICACION      0
COD_BARRIO              0
BARRIO                  0
COD_DISTRITO            0
DISTRITO                0
ESTADO                  0
COORD_GIS_X             0
COORD_GIS_Y             0
SISTEMA_COORD           0
LATITUD                 0
LONGITUD                0
TIPO_VIA               97
NOM_VIA                97
NUM_VIA                98
COD_POSTAL             37
DIRECCION_AUX         243
NDP                   126
FECHA_INSTALACION     114
CODIGO_INTERNO         69
CONTRATO_COD            0
TIPO                    0
TOTAL_ELEM              0
dtype: int64


A través de la función isnull, podemos observar la cantidad de valores nulos que hay en cada una de las columnas.

# 6. Eliminación de valores nulos

In [11]:
datos_sin_nulos = df_rawdata.dropna(inplace=True)

Una vez visto los valores nulos que hay en cada una de las columnas, decido eliminar del Dataframe original todas aquellas filas que contienen valores nulos. Al ser un Dataframe con bastante datos, considero que no afectará a la posterior interpretación de los datos.

# 7. Comprobación de la eliminación de los valores nulos

In [12]:
valores_nulos=df_rawdata.isnull().sum()
print(valores_nulos)

ID                    0
DESC_CLASIFICACION    0
COD_BARRIO            0
BARRIO                0
COD_DISTRITO          0
DISTRITO              0
ESTADO                0
COORD_GIS_X           0
COORD_GIS_Y           0
SISTEMA_COORD         0
LATITUD               0
LONGITUD              0
TIPO_VIA              0
NOM_VIA               0
NUM_VIA               0
COD_POSTAL            0
DIRECCION_AUX         0
NDP                   0
FECHA_INSTALACION     0
CODIGO_INTERNO        0
CONTRATO_COD          0
TIPO                  0
TOTAL_ELEM            0
dtype: int64


# 8. Veamos como queda el Dataframe resumido tras la eliminación de los valores nulos:

In [13]:
df_rawdata.shape

(1883, 23)

Del total de 2162 filas, nos quedamos con 1883 filas que contienen datos con valores no nulos.

# 9. Cambio de los formatos de las columnas COD_POSTAL y FECHA_INSTALACIÓN:

In [14]:
df_rawdata['COD_POSTAL'] = df_rawdata['COD_POSTAL'].astype(str).str.rstrip('.0')
print(df_rawdata['COD_POSTAL'])

1       28035
4       28032
7       28011
8        2805
9       28012
        ...  
2155    28035
2156    28042
2157    28041
2158    28042
2160    28025
Name: COD_POSTAL, Length: 1883, dtype: object


In [15]:
df_rawdata['FECHA_INSTALACION'] = pd.to_datetime(df_rawdata['FECHA_INSTALACION'])
df_rawdata['FECHA_INSTALACION'] = df_rawdata['FECHA_INSTALACION'].dt.date
print(df_rawdata['FECHA_INSTALACION'])

1       2007-01-09
4       2003-01-01
7       2006-06-19
8       2002-01-01
9       2007-05-17
           ...    
2155    2006-04-18
2156    2007-01-01
2157    2004-01-01
2158    2006-01-01
2160    2004-01-01
Name: FECHA_INSTALACION, Length: 1883, dtype: object


Con la función astype y to.datetime, hemos cambiado el formato de las columnas 'COD_POSTAL' y 'FECHA_INSTALACIÓN' para que los datos que se muestren en estas columnas se puedan ver de manera adecuada. A través de está homgeneización de los datos, podemos observar que, en la columna 'COD_POSTAL' todavía hay datos que no están del todo correctos. Códigos postales con 4 dígitos en vez de 5.

# 10. Filtrado de datos en la columna 'COD_POSTAL':

In [16]:
filas_incorrectas = df_rawdata[df_rawdata['COD_POSTAL'].astype(str).str.len() !=5]
print(filas_incorrectas.shape)

(145, 23)


A través del filtrado, podemos observar el número de filas que no contienen 5 dígitos en la columna de 'COD_POSTAL'. Al ser un valor no elevado, decido eliminar estas filas del Dataframe original.

# 11. Eliminación de datos en columna 'COD_POSTAL'

In [17]:
df_rawdata = df_rawdata[df_rawdata['COD_POSTAL'].astype(str).str.len() == 5].reset_index(drop=True)
print(df_rawdata.shape)


(1738, 23)


A hora podemos ver que las filas que no contenian 5 dígitos en la columna 'COD_POSTAL' han sido eliminadas del Dataframe original.

# 12. Filtrado de datos

In [18]:
print(df_rawdata['ESTADO'].unique())

['OPERATIVO']


Para explorar la viabilidad de filtrar los datos en base a la columna 'ESTADO' utilizamos el método unique. Con este código vemos que esta variable no nos sirve para poder hacer el filtrado, por lo tanto, buscamos otra que si sirva.

# 13. Filtrado por 'COD_BARRIO' 101

In [19]:
filtrado_barrio_101 = df_rawdata[df_rawdata['COD_BARRIO'] ==101]
nuevo_dataframe = filtrado_barrio_101
print(nuevo_dataframe)

         ID DESC_CLASIFICACION  COD_BARRIO        BARRIO  COD_DISTRITO  \
301   17330      Área Infantil         101  LOS CARMENES            10   
343   17355      Área Infantil         101  LOS CARMENES            10   
385   17360      Área Infantil         101  LOS CARMENES            10   
413   17341      Área Infantil         101  LOS CARMENES            10   
463   17329      Área Infantil         101  LOS CARMENES            10   
540   17352      Área Infantil         101  LOS CARMENES            10   
632   17344      Área Infantil         101  LOS CARMENES            10   
695   17331      Área Infantil         101  LOS CARMENES            10   
811   17356      Área Infantil         101  LOS CARMENES            10   
949   17342      Área Infantil         101  LOS CARMENES            10   
1022  17385      Área Infantil         101  LOS CARMENES            10   
1042  17394      Área Infantil         101  LOS CARMENES            10   
1287  17365      Área Infantil        

Tras realizar el filtro por el código del barrio observamos que el código 101 se trata del distrinto la LATINA y a en este distrito se ha realizado 13 instalaciones de áreas infantiles entre los años 2004 y 2007.

# 14. Suma de la columna 'TOTAL_ELEM':

In [20]:
nuevo_dataframe = df_rawdata[df_rawdata['COD_BARRIO'] == 101].copy()
nuevo_dataframe['TOTAL'] = nuevo_dataframe['TOTAL_ELEM'].cumsum()
print(nuevo_dataframe)

         ID DESC_CLASIFICACION  COD_BARRIO        BARRIO  COD_DISTRITO  \
301   17330      Área Infantil         101  LOS CARMENES            10   
343   17355      Área Infantil         101  LOS CARMENES            10   
385   17360      Área Infantil         101  LOS CARMENES            10   
413   17341      Área Infantil         101  LOS CARMENES            10   
463   17329      Área Infantil         101  LOS CARMENES            10   
540   17352      Área Infantil         101  LOS CARMENES            10   
632   17344      Área Infantil         101  LOS CARMENES            10   
695   17331      Área Infantil         101  LOS CARMENES            10   
811   17356      Área Infantil         101  LOS CARMENES            10   
949   17342      Área Infantil         101  LOS CARMENES            10   
1022  17385      Área Infantil         101  LOS CARMENES            10   
1042  17394      Área Infantil         101  LOS CARMENES            10   
1287  17365      Área Infantil        

Sobre el filtrado realizado en la columna 'COD_BARRIO', hacemos una suma del total de elementos que se han instalado. El resultado son 130 elementos instalados en el barrio de la Latina en el periodo de 2004 y 2007.

# 15. Ordenar el nuevo Dataframe en base a la fecha de instalación ('FECHA_INSTALACION'), de más antiguo a más reciente:  

In [21]:
nuevo_dataframe = df_rawdata[df_rawdata['COD_BARRIO'] == 101].copy()
nuevo_dataframe['FECHA_INSTALACION'] = pd.to_datetime(nuevo_dataframe['FECHA_INSTALACION'])
nuevo_dataframe.sort_values(by='FECHA_INSTALACION', inplace=True)
nuevo_dataframe['TOTAL'] = nuevo_dataframe['TOTAL_ELEM'].cumsum()
print(nuevo_dataframe)

         ID DESC_CLASIFICACION  COD_BARRIO        BARRIO  COD_DISTRITO  \
632   17344      Área Infantil         101  LOS CARMENES            10   
949   17342      Área Infantil         101  LOS CARMENES            10   
695   17331      Área Infantil         101  LOS CARMENES            10   
463   17329      Área Infantil         101  LOS CARMENES            10   
385   17360      Área Infantil         101  LOS CARMENES            10   
1287  17365      Área Infantil         101  LOS CARMENES            10   
413   17341      Área Infantil         101  LOS CARMENES            10   
301   17330      Área Infantil         101  LOS CARMENES            10   
1022  17385      Área Infantil         101  LOS CARMENES            10   
343   17355      Área Infantil         101  LOS CARMENES            10   
540   17352      Área Infantil         101  LOS CARMENES            10   
811   17356      Área Infantil         101  LOS CARMENES            10   
1042  17394      Área Infantil        

Ahora tenemos el Dataframe filtrado por el distrito 101 (corrspondiente al barrio la Latina), con la fecha de instalción ordenada de más antiguo a más reciente y con la nueva columna del total de elementos instalados en las áreas infantiles de este barrio.

# 16. Agregación con el método groupby

In [64]:
print(nuevo_dataframe.groupby('FECHA_INSTALACION')['TOTAL_ELEM'].size())

FECHA_INSTALACION
2004-12-10    2
2005-01-12    1
2005-01-14    1
2005-01-17    1
2005-01-24    1
2005-04-28    1
2006-06-19    1
2006-06-21    1
2007-02-13    2
2007-02-20    1
2007-04-10    1
Name: TOTAL_ELEM, dtype: int64


Al tratar de agrupar las columnas 'FECHA_INSTALACION' y 'TOTAL_ELEM'observamos la agrupación no muestra de manera correcta el total de elementos instalados por año en el distrito 101, barrio la Latina. 

# 17. Método Groupby en base al año de instalación de los elementos en las áreas infantiles del barrio la Latina.

In [67]:
nuevo_dataframe['FECHA_INSTALACION'] = pd.to_datetime(nuevo_dataframe['FECHA_INSTALACION'])
nuevo_dataframe['AÑO_INSTALACION'] = nuevo_dataframe['FECHA_INSTALACION'].dt.year
print(nuevo_dataframe.groupby('AÑO_INSTALACION')['TOTAL_ELEM'].sum())

AÑO_INSTALACION
2004    13
2005    54
2006     9
2007    54
Name: TOTAL_ELEM, dtype: int64


Tras haber extraído previamente el año de la columna 'FECHA_INSTALACION' y haber creado una nueva columna de 'AÑO_INSTALACION', se ha podido emplear correctamente el método groupby, sin haber almacenado ninguna variable correspodiente al método groupby. El resultado final, nos muestra el total de elementos que se han instalado en el barrio de la Latina por años, del periodo del 2004 al 2007.

## Evaluación

1.	**Calificación:** 2,5 / 2,5  
    **Comentarios:**  
2.	**Calificación:** 2,5 / 2,5  
    **Comentarios:**  
3.	**Calificación:** 2,5 / 2,5    
    **Comentarios:**    
4.	**Calificación:** 2,5 / 2,5  
    **Comentarios:**  

### Total: *10 / 10*